docs/decisions/infra/0009-separate-app-infrastructure-into-layers.md

```diff
@@ -0,0 +1,40 @@
+# Separate app infrastructure into layeres
+
+* Status: accepted
+* Deciders: @lorenyu @rocketnova @jamesbursa
+* Date: 2023-09-11
+
+Technical Story: [Document rationale for splitting up infra layers across multiple root modules](https://github.com/navapbc/template-infra/issues/431)
+
+## Context and Problem Statement
+
+This document builds on the database module design [ADR: Separate the database infrastructure into a separate layer](./0005-separate-database-infrastructure-into-separate-layer.md) to describe the general rationale for separating application environment infrastructure into separate root modules that are managed in separate terraform state files and updated separately rather than all in one single root module. It restates and summarizes the rationale from the previous ADR and includes additional motivating examples.
+
+## Overview
+
+Based on the factors in the section below, the infrastructure has been groups into the following separate layers:
+
+* Account layer
+* Network layer
+* Build repository layer
```

Comment:
Non-blocking clarifying question: If a project has multiple apps that live in the service layer, do you envision that they share the same build repository? Or would they each have their own?

Reply:
Oops I already merged before I saw this. I'll answer after my meeting.

Reply:
No worries! Non-blocking question that I added after I approved, mostly for my own understanding.

Reply:
> If a project has multiple apps that live in the service layer, do you envision that they share the same build repository? Or would they each have their own?

Each application would have it's own build repository. The principles is that each "app" should be treated as an independently deployed system component. For example, each app should be treated like a (micro)service, and the only way apps can talk to each other is via API calls over the network. In that situation, each app wouldn't be able to access infra resources that belong to other apps (e.g. database).

For example, a monorepo with two apps `web-app` and `static-app` might have a folder structure that looks like this under `infra/`

* web-app
  * build-repository // container image repository using ECR
  * database
  * service // ALB + ECS service
* static-app
  * build-repository // static site builds stored in S3
  * service // cloudfront + S3 bucket

the "static-app" example is included as a vision of what it'd be like if/when we include support for static sites.

Reply:
That makes sense. Based on this documentation, `web-app` and `static-app` could share the same account. Would it also fit within the paradigm for them to share the same network (aka be within the same VPC)?

Reply:
Yeah `web-app` and `static-app` could share the same account or be in different accounts. The mapping of app/environment to account is in [account_names_by_environment in app-config](https://github.com/navapbc/template-infra/blob/main/infra/app/app-config/main.tf#L19-L54). the build repository lives in the "shared" environment.

re: VPC, yeah they can also be in the same VPC and use the same WAF, or be in separate VPCs. the VPC part isn't developed yet. I have a [milestone (epic) for it](https://github.com/navapbc/template-infra/milestone/4). Of course, if the two apps are in the same VPC, then they must be in the same account as well.

Reply:
Great. Thank you!